File: week13/11/5/P5.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        return help(root, root->val, root->val);
    }

    int help(TreeNode* cur, int big, int small){
        if (!cur){
            return big - small;
        }
        int b = max(big, cur->val);
        int s = min(small, cur->val);

        int ld = help(cur->left, b, s);
        int rd = help(cur->right, b, s);

        return max(ld, rd);
    }
};

File: week13/11/5/P5.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        return self.__DFTHelp(root, root.val, root.val)

    def __DFTHelp(self, root, big, small) -> int:
        if not root:
            return big - small
        big = max(root.val, big)
        small = min(root.val, small)
        
        #big, small에 현재 root val이 포함이 되기 때문에 left랑 right만 비교하면 됨
        ld = self.__DFTHelp(root.left, big, small)
        rd = self.__DFTHelp(root.right, big, small)

        return max(ld,rd)